elide-core/src/test/java/com/yahoo/elide/core/LifeCycleTest.java

```diff
@@ -395,6 +428,130 @@ public void testUpdateWithChangeSpec() {
+        verify(onUpdateDeferredCallback, times(2)).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdateImmediateCallback, times(2)).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitCallback, times(2)).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitAuthor, never()).execute(any(), isA(RequestScope.class), any());
+
+        verifyNoEmptyCallbacks();
+    }
+
+    @Test
+    public void testUpdateRelationshipWithChangeSpec() {
+        Book book = new Book();
+        Author author = new Author();
+        book.setAuthors(Sets.newHashSet(author));
+        author.setBooks(Sets.newHashSet(book));
+        DataStoreTransaction tx = mock(DataStoreTransaction.class);
+        when(tx.getRelation(any(), eq(author), eq("books"), any(), any(), any(), any())).then((i) -> author.getBooks());
+        when(tx.getRelation(any(), eq(book), eq("authors"), any(), any(), any(), any())).then((i) -> book.getAuthors());
+
+        RequestScope scope = new RequestScope(null, null, tx , new User(1), null, getElideSettings(null, dictionary, MOCK_AUDIT_LOGGER),
+                false);
+        PersistentResource<Author> resourceBook = new PersistentResource(book, null, scope.getUUIDFor(book), scope);
+        PersistentResource<Author> resourceAuthor = new PersistentResource(author, null, scope.getUUIDFor(book), scope);
+
+        verify(onUpdateDeferredCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdateImmediateCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitAuthor, never()).execute(eq(book), isA(RequestScope.class), any());
+
+        //Verify changeSpec is passed to hooks
+        resourceAuthor.removeRelation("books", resourceBook);
+
+        scope.runQueuedPreSecurityTriggers();
+        verify(onUpdatePostCommitCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitAuthor, never()).execute(eq(book), isA(RequestScope.class), any());
+
+        scope.runQueuedPreCommitTriggers();
+        verify(onUpdatePostCommitCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitAuthor, never()).execute(eq(book), isA(RequestScope.class), any());
+
+        scope.getPermissionExecutor().executeCommitChecks();
+        verify(onUpdatePostCommitCallback, never()).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdatePostCommitAuthor, never()).execute(eq(book), isA(RequestScope.class), any());
+
+        scope.runQueuedPostCommitTriggers();
+        verify(onUpdateDeferredCallback, times(1)).execute(eq(book), isA(RequestScope.class), any());
+        verify(onUpdateImmediateCallback, times(1)).execute(eq(book), isA(RequestScope.class), any());
```

Comment:
Should we verify these above as well (immediately after it would trigger)?

Reply:
Sure, but we only need to verify the trigger only after it happens, and again at the end of the function to make sure there were no extra calls.